swarm/src/dial_opts.rs

```diff
@@ -297,29 +297,34 @@ impl WithoutPeerIdWithAddress {
     }
 }
 
-/// The available conditions under which a new dialing attempt to
-/// a known peer is initiated.
-///
-/// ```
-/// # use libp2p_swarm::dial_opts::{DialOpts, PeerCondition};
-/// # use libp2p_identity::PeerId;
-/// #
-/// DialOpts::peer_id(PeerId::random())
-///    .condition(PeerCondition::Disconnected)
-///    .build();
-/// ```
-#[derive(Debug, Copy, Clone, Default)]
-pub enum PeerCondition {
-    /// A new dialing attempt is initiated _only if_ the peer is currently
-    /// considered disconnected, i.e. there is no established connection
-    /// and no ongoing dialing attempt.
-    #[default]
-    Disconnected,
-    /// A new dialing attempt is initiated _only if_ there is currently
-    /// no ongoing dialing attempt, i.e. the peer is either considered
-    /// disconnected or connected but without an ongoing dialing attempt.
-    NotDialing,
-    /// A new dialing attempt is always initiated, only subject to the
-    /// configured connection limits.
-    Always,
+bitflags::bitflags! {
+    /// The available conditions under which a new dialing attempt to
+    /// a known peer is initiated. These conditions can be combined.
+    ///
+    /// ```
+    /// # use libp2p_swarm::dial_opts::{DialOpts, PeerCondition};
+    /// # use libp2p_identity::PeerId;
+    /// #
+    /// DialOpts::peer_id(PeerId::random())
+    ///    .condition(PeerCondition::Disconnected | PeerCondition::NotDialing)
+    ///    .build();
+    /// ```
+    #[derive(Debug, Copy, Clone)]
+    pub struct PeerCondition: u32 {
```

Comment:
The fact that this changes from an enum to a struct is a breaking change anyway.

My suggestion would be to leave `PeerCondition` as it is and introduce a new `DialCondition` bitflag.

Then we can deprecate `DialOpts::peer_condition` and introduce `DialOpts::dial_condition`. That is backwards-compatible and allows us to ship this earlier :)

The default value is tricky. I think we should have the same default (`Disconnected`) but we must only evaluate that if the user did not explicitly set a `PeerCondition`. Does that make sense? We should be able to capture that with an additional boolean inside the `peer_condition` function.